Add watcher-specific CI tests

tt-dma opened the pull request:
- Refactor dispatch-agnostic test fixtures to have a common fixture, and dprint/watcher-specific fixtures
- Add tests for watcher waypointing and sanitization
- Some minor changes tot he watcher server to enable gtesting

tt-dma committed:
#4756: Rework dispatch-agnostic gtest fixtures to allow for more types

tt-dma committed:
#4756: Add a watcher CI test for waypointing feature

tt-dma committed:
#4756: Add a watcher noc sanitization gtest, as well as a way to detect
a sanitization failure from outside the watcher server.

tt-dma commented on tests/tt_metal/tt_metal/unit_tests_common/watcher/test_noc_sanitize.cpp:
```diff
@@ -0,0 +1,125 @@
+    // Skip this test for slow dipatch for now. Due to how llrt currently sits below device, it's
+    // tricky to check watcher server status from the finish loop for slow dispatch. Once issue #4363
+    // is resolved, we should add a check for print server handing in slow dispatch as well.
+    if (this->slow_dispatch_)
```
Issue here is that fast dispatch is not supported yet for the remote device. This means that the test does not currently check the functionality on remote device. Plan is to leave this as-is for this PR, since we are actively bringing up fast dispatch for remote device.

tt-dma replied:
I have manually verified that this feature does work for the remote device

pgkeller commented on tt_metal/llrt/watcher.hpp:
```diff
@@ -20,6 +20,13 @@ void watcher_sanitize_host_noc_write(const metal_SocDescriptor &soc_d, CoreCoord
+// Check whether the watcher server has been killed due to an error detected.
+bool watcher_server_killed_due_to_error();
+// Function to clear this flag, so that non-watcher runs can continue as normal.
+// TODO(dma): this doesn't currently clear the actual error codes on the defice. Once watcher is
```
typo defice

tt-dma replied:
Fixed this in latest commit

pgkeller commented on tt_metal/llrt/watcher.cpp:
```diff
@@ -495,7 +498,13 @@ static void watcher_loop(int sleep_usecs) {
+                watcher::watcher_killed_due_to_error = true;
+                watcher::enabled = false;
+                break;
+            }
```
Is the flow now that throw's only occur in wait_finish in the queue?  I don't think that'll work for all cases, it is possible a program could enqueue work for a long time w/o ever calling finish (that's not the case in our tests).

tt-dma replied:
Updated this as we discussed, I've manually verified that the exception throws instantly if the setting is disabled (which is default).

pgkeller commented on tests/tt_metal/tt_metal/test_kernels/misc/watcher_waypoints.cpp:
```diff
@@ -0,0 +1,50 @@
+#include "debug/status.h"
+
+// Wait cycles to create a ~1sec delay
+#define WAIT_CYCLES 1200000000
```
I know I'm contradicting myself, but we can grab this on the host through this API:
int tt_SiliconDevice::get_clock(int logical_device_id)
Almeet may know how to plumb this

tt-dma replied:
Fixed this in latest commit

pgkeller commented on tests/tt_metal/tt_metal/test_kernels/misc/watcher_waypoints.cpp:
```diff
@@ -0,0 +1,50 @@
+// Wait cycles to create a ~1sec delay
+#define WAIT_CYCLES 1200000000
+// Flag address for syncing between cores (BRISC controls this)
+#define SYNC_ADDR 409600
```
Again, contradicting myself.  We could allocate an L1 buffer then pass the address in from the host either w/ a define or an argument

tt-dma replied:
Fixed this in latest commit

pgkeller commented on tests/tt_metal/tt_metal/unit_tests_common/common/watcher_fixture.hpp:
```diff
@@ -0,0 +1,64 @@
+        std::remove(log_file_name.c_str());
+
+        // Reset watcher settings
+        tt::llrt::OptionsG.set_watcher_enabled(false);
```
thinking we should query the initial state and restore it in case someone is running a bunch of tests w/ watcher enabled.

tt-dma replied:
Fixed this in latest commit

tt-dma committed:
#4756: Fix watcher waypoint test to query clock frequency and allocate a
L1 buffer for syncing

Also update watcher test fixture to restore previous state

tt-dma committed:
#4756: Add a hidden option to catch/throw watcher&dprint exceptions

No environment variable for this, default is to throw so that errors are
shown to the user right away. Can be set to "catch", where
errors are logged to be checked by tests.

pgkeller commented on tt_metal/llrt/rtoptions.hpp:
```diff
@@ -56,6 +58,14 @@ class RunTimeOptions {
     inline int get_watcher_append()                   { return watcher_append; }
     inline void set_watcher_append(bool append)       { watcher_append = append; }
+    inline bool get_watcher_catch_exceptions()        { return watcher_catch_exceptions; }
+    inline void set_watcher_catch_exceptions(bool catch_exceptions) {
```
thinking we should collapse the 2 into 1: ie, we're either in "test mode" in which case we catch exceptions or we're in "run mode" in which case we throw exceptions.  having this per-concept is likely overkill (and possibly more maintenance)

tt-dma replied:
Good point, changed this to a single setting in latest rebase

DrJessop commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -1071,6 +1072,12 @@ void CommandQueue::wait_finish() {
             TT_THROW("Command Queue could not finish: device hang due to unanswered DPRINT WAIT.");
         }
+
+        // If the watcher has detected a sanitization error, the server will have closed and and a
```
typo (double 'and')

tt-dma replied:
Fixed in latest rebase!

tt-dma commented:
CI passing: https://github.com/tenstorrent-metal/tt-metal/actions/runs/7661110404
